soroban-sdk/src/account.rs

```diff
@@ -7,9 +7,169 @@ use core::{
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_into_val(self, env: &Env) -> Result<AccountId, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(env, self)
+    }
+}
+
+impl TryFromVal<Env, RawVal> for AccountId {
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_from_val(env: &Env, val: RawVal) -> Result<Self, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(env, val.try_into()?)
+    }
+}
+
+impl TryIntoVal<Env, AccountId> for RawVal {
+    type Error = <Account as TryFromVal<Env, Object>>::Error;
+
+    fn try_into_val(self, env: &Env) -> Result<AccountId, Self::Error> {
+        <_ as TryFromVal<_, RawVal>>::try_from_val(env, self)
+    }
+}
+
+impl IntoVal<Env, Object> for AccountId {
+    fn into_val(self, _env: &Env) -> Object {
+        self.to_object()
+    }
+}
+
+impl IntoVal<Env, Object> for &AccountId {
+    fn into_val(self, _env: &Env) -> Object {
+        self.to_object()
+    }
+}
+
+impl IntoVal<Env, RawVal> for AccountId {
+    fn into_val(self, _env: &Env) -> RawVal {
+        self.to_raw()
+    }
+}
+
+impl IntoVal<Env, RawVal> for &AccountId {
+    fn into_val(self, _env: &Env) -> RawVal {
+        self.to_raw()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+use super::xdr::ScVal;
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<&AccountId> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: &AccountId) -> Result<Self, Self::Error> {
+        ScVal::try_from_val(&v.0.env, v.0.val.to_raw())
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFrom<AccountId> for ScVal {
+    type Error = ConversionError;
+    fn try_from(v: AccountId) -> Result<Self, Self::Error> {
+        (&v).try_into()
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for AccountId {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: ScVal) -> Result<Self, Self::Error> {
+        <_ as TryFromVal<_, Object>>::try_from_val(
+            env,
+            val.try_into_val(env).map_err(|_| ConversionError)?,
+        )
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryIntoVal<Env, AccountId> for ScVal {
+    type Error = ConversionError;
+    fn try_into_val(self, env: &Env) -> Result<AccountId, Self::Error> {
+        AccountId::try_from_val(env, self)
+    }
+}
+
+impl AccountId {
+    pub(crate) unsafe fn unchecked_new(obj: EnvObj) -> Self {
+        Self(obj)
+    }
+
+    pub fn env(&self) -> &Env {
+        self.0.env()
+    }
+
+    pub fn as_raw(&self) -> &RawVal {
+        self.0.as_raw()
+    }
+
+    pub fn as_object(&self) -> &Object {
+        self.0.as_object()
+    }
+
+    pub fn to_raw(&self) -> RawVal {
+        self.0.to_raw()
+    }
+
+    pub fn to_object(&self) -> Object {
+        self.0.to_object()
+    }
+}
+
 /// Account references a Stellar account and provides access to information
 /// about the account, such as its thresholds and signers.
 #[derive(Clone)]
```

Comment:
Note that the Account type, and its functions are still accessed via a `BytesN<32>`. That will be changed to use the new `AccountId` in https://github.com/stellar/rs-soroban-env/issues/477.